src/GoogleAuthManager.php

```diff
@@ -3,93 +3,193 @@
+   * @param \Drupal\Core\Logger\LoggerChannelFactoryInterface $logger_factory
+   *   Used for logging errors.
+   * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface $event_dispatcher
+   *   Used for dispatching events to other modules.
+   * @param \Drupal\Core\Entity\EntityFieldManagerInterface $entity_field_manager
+   *   Used for accessing Drupal user picture preferences.
+   * @param \Drupal\Core\Routing\UrlGeneratorInterface $url_generator
+   *   Used for generating absoulute URLs.
    */
-  public function __construct(RequestStack $request) {
-    $this->request = $request->getCurrentRequest();
+  public function __construct(LoggerChannelFactoryInterface $logger_factory, EventDispatcherInterface $event_dispatcher, EntityFieldManagerInterface $entity_field_manager, UrlGeneratorInterface $url_generator) {
+    $this->loggerFactory      = $logger_factory;
+    $this->eventDispatcher    = $event_dispatcher;
+    $this->entityFieldManager = $entity_field_manager;
+    $this->urlGenerator       = $url_generator;
   }
 
   /**
-   * {@inheritdoc}
+   * Authenticates the users by using the access token.
+   *
+   * @return $this
+   *   The current object.
    */
   public function authenticate() {
-    $this->client->setAccessToken($this->getAccessToken());
-    return $this;
+    $this->token = $this->client->getAccessToken('authorization_code',
+      ['code' => $_GET['code']]);
+
+    return $this->token;
   }
 
   /**
-   * {@inheritdoc}
+   * Gets the data by using the access token returned.
+   *
+   * @return array
+   *   User info returned by the Google.
    */
-  public function getAccessToken() {
-    if (!$this->accessToken) {
-      $this->accessToken = $this->client->fetchAccessTokenWithAuthCode($this->getCode());
-    }
-
-    return $this->accessToken;
+  public function getUserInfo() {
+    $this->user = $this->client->getResourceOwner($this->token);
+    return $this->user;
   }
 
   /**
-   * {@inheritdoc}
+   * Returns the Google login URL where user will be redirected.
+   *
+   * @return string
+   *   Absolute Google login URL where user will be redirected
    */
-  public function getUserInfo() {
-    return $this->getOauth2Service()->userinfo->get();
+  public function getGoogleLoginUrl($data_points) {
+    $scopes = $this->checkForScopes($data_points);
+    $login_url = $this->client->getAuthorizationUrl($scopes);
+
+    // Generate and return the URL where we should redirect the user.
+    return $login_url;
   }
 
   /**
-   * Gets Google Oauth2 Service.
+   * Returns scopes required for data point defined by administator.
+   *
+   * @param array $data_points
+   *   The data points to be collected.
    *
-   * @return Google_Service_Oauth2
-   *   The Google Oauth2 service.
+   * @return array
+   *   scopes for authorization URL.
    */
-  protected function getOauth2Service() {
-    return new Google_Service_Oauth2($this->getClient());
+  protected function checkForScopes($data_points) {
```

Comment:
I don't understand why this method is necessary. It seems to me it only allows 'name' and 'email'